### PR TITLE
Micrometer Observation instrumentation

### PR DESCRIPTION
**Pull request by marcingrzejszczak**
fixes https://github.com/line/armeria/issues/4659

### CODE REVIEW

File: observation/src/main/java/com/linecorp/armeria/client/observation/HttpClientObservationConvention.java
Comment (jrhee17) [inline observation/src/main/java/com/linecorp/armeria/client/observation/HttpClientObservationConvention.java:22]: Question) Is it possible to remove this interface and just add `supportContext` directly into `DefaultHttpClientObservationConvention`?

File: observation/src/main/java/com/linecorp/armeria/client/observation/HttpClientContext.java
Comment (jrhee17) [inline observation/src/main/java/com/linecorp/armeria/client/observation/HttpClientContext.java:47]: I believe `uri` is non-null so we can get rid of the null checks here and at L55

**Review by jrhee17 — COMMENTED** [review]
Sorry about the delay! Left some questions and comments but honestly I think this PR is already close to completion 🚀 

Also pushed some small commits which
1. Fixes failing tests
2. Hides some public classes

Feel free to revert any changes I made, or just let me know and I'll revert it back 😄

File: observation/src/main/java/com/linecorp/armeria/client/observation/HttpClientObservationDocumentation.java
Comment (jrhee17) [reply observation/src/main/java/com/linecorp/armeria/client/observation/HttpClientObservationDocumentation.java:33]: I guess `SESSION_PROTOCOL` and `SERIALIZATION_FORMAT` can also be considered low cardinality

File: observation/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java
Comment (jrhee17) [inline observation/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java:90]: Nice 👍

File: observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java
Comment (jrhee17) [inline observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java:129]: I think this information isn't really critical to merge this PR, but still is useful for users going through traces.

Would it be difficult to also provide the option to pass a `timestamp` such that  `observation.event(Event, long)` like brave and otel do? (maybe `DefaultMeterObservationHandler` can just ignore the timestamp)

**Review by marcingrzejszczak — COMMENTED** [review]
I've added comments to your comments and will upload some code very soon

File: observation/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java
Comment (marcingrzejszczak) [inline observation/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java:72]: So I understand that you're not allowing users to modify the default tagging by injecting the custom convention?

File: site/src/pages/docs/advanced-zipkin.mdx
Comment (jrhee17) [inline site/src/pages/docs/advanced-zipkin.mdx:50]: Can you revert this since the class doesn't exist? 😅 
I'll be adding a separate page on tracing later

**Review by jrhee17 — COMMENTED** [review]
I'll take another look in the morning but I think this pull request is pretty much done!

I've pushed some changes which
- Makes `HttpClientContext`, `HttpServerContext` public along with some test code
  - Also renamed some variables
- Removed the default `Http[Client|Server]ObservationConvention` interfaces
- Added some more documentation

Let me know if any changes don't make sense 😄 
Also, what do you think of changing the status of this PR to ready for review now?

**Comment by marcingrzejszczak** [comment]
OK, I've removed setting of the remote address - and applied all suggested changes. Also, the PR is now officially ready for review :)

File: observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java
Comment (jrhee17) [inline observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java:125]:
```suggestion
```

**Review by jrhee17 — APPROVED** [review]
Thanks @marcingrzejszczak 🙇 👍 🚀

**Comment by marcingrzejszczak** [comment]
Thank you, my pleasure @jrhee17 ! :)

File: observation/src/test/java/com/linecorp/armeria/server/observation/MicrometerObservationServiceTest.java
Comment (trustin) [inline observation/src/test/java/com/linecorp/armeria/server/observation/MicrometerObservationServiceTest.java:90]: @jrhee17 Could you undisable and fix it?

File: settings.gradle
Comment (trustin) [inline settings.gradle:66]: Should we keep this as a separate module or make it part of the core? We already have Micrometer in our core, so I think we can move it into core if it doesn't pull in other dependencies such as Brave.

File: observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java
Comment (marcingrzejszczak) [reply observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java:142]: No there is none. That's why this (https://github.com/line/armeria/pull/4980#discussion_r1252821558) is not possible to be done

File: settings.gradle
Comment (marcingrzejszczak) [reply settings.gradle:66]: Makes sense. We always release Micrometer Observation with Micrometer Core (that's the same project) and Core depends on Observation.

**Comment by marcingrzejszczak** [comment]
Is there anything else I should to merge this?

**Comment by jrhee17** [comment]
Sorry about the delay 😅    Let me update the PR to address @trustin's comments tonight, and then ping the other maintainers

**Comment by marcingrzejszczak** [comment]
Oh absolutely no problem, I was just wondering if someone is waiting on me and I have missed the message. Take your time!

File: observation/src/test/java/com/linecorp/armeria/server/observation/MicrometerObservationServiceTest.java
Comment (jrhee17) [reply observation/src/test/java/com/linecorp/armeria/server/observation/MicrometerObservationServiceTest.java:90]: Let me remove this test 😅 

Essentially the original test was checking if `BraveClient` threw an exception if not used with `RequestContextCurrentTraceContext`.

In our current setup, we actually encourage usage of brave's default `ThreadLocalCurrentTraceContext` (although we could still use `RequestContextCurrentTraceContext`)

If we want to enable this, this would mean we need to migrate over this validation which I didn't think makes much sense. Let me know if you feel otherwise though.

https://github.com/line/armeria/blob/262fb4f0dd70a6c34b1c900b3a9156edc62cfcb1/brave/src/main/java/com/linecorp/armeria/internal/common/brave/TraceContextUtil.java#L62

File: settings.gradle
Comment (jrhee17) [reply settings.gradle:66]: Talked with the team and moving the implementation to the `:core` module

File: core/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java:118]: Ditto. Should we skip observation if `observationRegistry.isNoop() == true || observation.isNoop() == true`?

**Review by ikhoon — COMMENTED** [review]
Overall looks good! Left comments for code style, micro optimizations, and suggestions.

File: core/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java
Comment (marcingrzejszczak) [reply core/src/main/java/com/linecorp/armeria/server/observation/MicrometerObservationService.java:118]: Yes, we can skip any instrumentation if registry or observation are noop

File: core/src/main/java/com/linecorp/armeria/client/observation/DefaultHttpClientObservationConvention.java
Comment (marcingrzejszczak) [reply core/src/main/java/com/linecorp/armeria/client/observation/DefaultHttpClientObservationConvention.java:35]: We could but I don't think that automated code generation will work with enums

File: core/src/main/java/com/linecorp/armeria/client/observation/DefaultHttpClientObservationConvention.java
Comment (marcingrzejszczak) [reply core/src/main/java/com/linecorp/armeria/client/observation/DefaultHttpClientObservationConvention.java:53]: Makes sense

Comment (marcingrzejszczak) [reply core/src/main/java/com/linecorp/armeria/client/observation/DefaultHttpClientObservationConvention.java:53]: So how would I extract the status from the `maybeGrpcTrailers`? Are those actually under `grpc-status` key?

**Comment by marcingrzejszczak** [comment]
Hey I think I've applied all the changes. Tell me if I missed sth

File: core/src/main/java/com/linecorp/armeria/client/observation/DefaultHttpClientObservationConvention.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/observation/DefaultHttpClientObservationConvention.java:142]:
```suggestion
        if (logAccess.isAvailable(RequestLogProperty.NAME)) {
```

File: core/src/main/java/com/linecorp/armeria/client/observation/ClientObservationContext.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/observation/ClientObservationContext.java:38]:
```suggestion
 *          .decorator(ObservationClient.newDecorator(registry, convention))
```

File: core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java:66]:
```suggestion
 *          .decorator(MicrometerObservationClient.newDecorator(observationRegistry))
```

File: core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java:142]: This condition is always false due to L121.
```suggestion
```

File: core/src/main/java/com/linecorp/armeria/server/observation/DefaultServiceObservationConvention.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/observation/DefaultServiceObservationConvention.java:143]:
```suggestion
        if (logAccess.isAvailable(RequestLogProperty.NAME)) {
```

File: core/src/main/java/com/linecorp/armeria/server/observation/ObservationService.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/observation/ObservationService.java:65]:
```suggestion
 *       .decorator(ObservationService.newDecorator(observationRegistry))
```

File: core/src/main/java/com/linecorp/armeria/server/observation/ObservationService.java
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/observation/ObservationService.java:143]:
```suggestion
```

**Review by ikhoon — APPROVED** [review]
High-quality work! Many thanks, @marcingrzejszczak. 🚀🙇‍♂️

**Comment by marcingrzejszczak** [comment]
Thank you! I've committed all of your suggestions. If there's anything else I should do I can do that in around 2 weeks when I'm back from pto

**Comment by marcingrzejszczak** [comment]
Hey I'm back from PTO. I've merged main into this branch. Is there anything else I should do?

**Comment by ikhoon** [comment]
Not yet. Other maintainers will also check this PR before merging this.

File: core/src/test/java/com/linecorp/armeria/client/observation/ObservationClientIntegrationTest.java
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/observation/ObservationClientIntegrationTest.java:144]: nit: It seems like we need to reindent this method.

File: core/src/test/java/com/linecorp/armeria/client/observation/ObservationClientIntegrationTest.java
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/observation/ObservationClientIntegrationTest.java:62]: `useHttp2Preface` is `true` by default so I think we can remove this `clientFactoryWithoutUpgradeRequest` completely.

File: core/src/test/java/com/linecorp/armeria/client/observation/ObservationClientIntegrationTest.java
Comment (minwoox) [inline core/src/test/java/com/linecorp/armeria/client/observation/ObservationClientIntegrationTest.java:101]:
```suggestion
```

File: core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java:117]: Question: Do the headers are added to the `newHeaders` when `start()` is called? (I have tried to find it, but I couldn't. 😓 )

**Review by minwoox — COMMENTED** [review]
I'm really sorry about the late review. 😓 
Looks great all in all. Left some nits and questions.
Thanks a lot for adding this nice feature! 🙇

**Review by minwoox — COMMENTED** [review]
I'm really sorry about the late review. 😓 
Looks great all in all. Left some nits and questions.
Thanks a lot for adding this nice feature! 🙇

**Review by minwoox — COMMENTED** [review]
I'm really sorry about the late review. 😓 
Looks great all in all. Left some nits and questions.
Thanks a lot for adding this nice feature! 🙇

**Comment by marcingrzejszczak** [comment]
@minwoox thank you for your review ! I applied all the suggested changes I think

**Comment by marcingrzejszczak** [comment]
I've rebased the branch against `main`

File: core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java
Comment (minwoox) [reply core/src/main/java/com/linecorp/armeria/client/observation/ObservationClient.java:120]: Ah ha got it. Thanks for the explanation. 👍

**Review by minwoox — APPROVED** [review]
Great! Thanks a lot! 😄 
Could you also fix the compile error? 😄

**Review by trustin — APPROVED** [review]
Thanks a lot for your patience, @marcingrzejszczak. Looks great to me. It'd be awesome if Micrometer Observation exposes an API that allows us to specify timestamps explicitly, though. Any chance we can have them in the future?

File: observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java
Comment (marcingrzejszczak) [reply observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java]: We could add that but the problem I see is that we can't set the timestamp for a counter. So whenever we have an event with Micrometer Observation what we do for metrics is we create a counter and for tracing we annotate the span. What we could do is ignore the timestamp information for the metrics and simply increment the counter when the event method was called. For tracing we would set an event with a given timestamp. Does it make sense? cc @shakuzen @jonatan-ivanov

Comment (trustin) [reply observation/src/main/java/com/linecorp/armeria/client/observation/MicrometerObservationClient.java]: > For tracing we would set an event with a given timestamp.
Yeah, this is probably what we want.

**Comment by trustin** [comment]
Thanks a lot for your high quality contribution and patience, @marcingrzejszczak! Let us stay tuned to https://github.com/micrometer-metrics/micrometer/issues/4032 :bow:

**Comment by marcingrzejszczak** [comment]
Sure @trustin, we had a conversation inside the team about this and we see no problem in adding that feature to Micrometer . Since I have your attention here, what do you think of https://github.com/netty/netty/issues/8546 ? :)

**Pull request merged**
